File: mobile/lib/models/disaster_report.dart
/// Disaster Report — data model for community-sourced disaster reports.
library;

/// Verification status assigned by Gemini LLM.
enum VerificationStatus { verified, rejected, flagged, pending }

/// Authority response status.
enum AuthorityStatus { none, dispatched, investigating, resolved }

/// Disaster category.
enum DisasterCategory { fire, flood, accident, infrastructure, medical, other }

class DisasterReport {
  const DisasterReport({
    required this.id,
    required this.userId,
    required this.lat,
    required this.lon,
    required this.imageUrl,
    required this.category,
    required this.severityScore,
    required this.llmConfidence,
    required this.verificationStatus,
    this.rejectionReason,
    this.description,
    this.authorityStatus = AuthorityStatus.none,
    required this.createdAt,
  });

  final String id;
  final String userId;
  final double lat;
  final double lon;
  final String imageUrl;
  final DisasterCategory category;
  final int severityScore;
  final double llmConfidence;
  final VerificationStatus verificationStatus;
  final String? rejectionReason;
  final String? description;
  final AuthorityStatus authorityStatus;
  final DateTime createdAt;

  factory DisasterReport.fromJson(Map<String, dynamic> json) {
    return DisasterReport(
      id: json['id'] as String,
      userId: json['userId'] as String? ?? '',
      lat: (json['lat'] as num).toDouble(),
      lon: (json['lon'] as num).toDouble(),
      imageUrl: json['imageUrl'] as String,
      category: _parseCategory(json['category'] as String?),
      severityScore: json['severityScore'] as int? ?? 3,
      llmConfidence: (json['llmConfidence'] as num?)?.toDouble() ?? 0,
      verificationStatus: _parseVerification(json['verificationStatus'] as String?),
      rejectionReason: json['rejectionReason'] as String?,
      description: json['description'] as String?,
      authorityStatus: _parseAuthority(json['authorityStatus'] as String?),
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }

  // ---------------------------------------------------------------------------
  // Helpers
  // ---------------------------------------------------------------------------

  String get categoryLabel => switch (category) {
    DisasterCategory.fire => 'Fire',
    DisasterCategory.flood => 'Flood',
    DisasterCategory.accident => 'Accident',
    DisasterCategory.infrastructure => 'Infrastructure',
    DisasterCategory.medical => 'Medical',
    DisasterCategory.other => 'Other',
  };

  String get severityLabel => switch (severityScore) {
    1 => 'Minor',
    2 => 'Moderate',
    3 => 'Significant',
    4 => 'Severe',
    5 => 'Critical',
    _ => 'Unknown',
  };

  String get statusLabel => switch (verificationStatus) {
    VerificationStatus.verified => 'Verified',
    VerificationStatus.rejected => 'Rejected',
    VerificationStatus.flagged => 'Flagged',
    VerificationStatus.pending => 'Pending',
  };

  String get authorityLabel => switch (authorityStatus) {
    AuthorityStatus.none => 'None',
    AuthorityStatus.dispatched => 'Dispatched',
    AuthorityStatus.investigating => 'Investigating',
    AuthorityStatus.resolved => 'Resolved',
  };

  static DisasterCategory _parseCategory(String? v) => switch (v) {
    'fire' => DisasterCategory.fire,
    'flood' => DisasterCategory.flood,
    'accident' => DisasterCategory.accident,
    'infrastructure' => DisasterCategory.infrastructure,
    'medical' => DisasterCategory.medical,
    _ => DisasterCategory.other,
  };

  static VerificationStatus _parseVerification(String? v) => switch (v) {
    'verified' => VerificationStatus.verified,
    'rejected' => VerificationStatus.rejected,
    'flagged' => VerificationStatus.flagged,
    _ => VerificationStatus.pending,
  };

  static AuthorityStatus _parseAuthority(String? v) => switch (v) {
    'dispatched' => AuthorityStatus.dispatched,
    'investigating' => AuthorityStatus.investigating,
    'resolved' => AuthorityStatus.resolved,
    _ => AuthorityStatus.none,
  };
}

/// Aggregated disaster stats from GET /disaster/stats.
class DisasterStats {
  const DisasterStats({
    required this.totalReports,
    required this.verified,
    required this.pending,
    required this.rejected,
    required this.flagged,
  });

  final int totalReports;
  final int verified;
  final int pending;
  final int rejected;
  final int flagged;

  factory DisasterStats.fromJson(Map<String, dynamic> json) {
    return DisasterStats(
      totalReports: json['totalReports'] as int? ?? 0,
      verified: json['verified'] as int? ?? 0,
      pending: json['pending'] as int? ?? 0,
      rejected: json['rejected'] as int? ?? 0,
      flagged: json['flagged'] as int? ?? 0,
    );
  }
}

File: backend/src/models/disaster-report.ts
/**
 * AfterMath Backend — Disaster Report model & Zod validation.
 *
 * Users submit GPS-pinned photo evidence of disasters. An LLM verifies
 * the image before the report is stored.
 */

import { z } from 'zod';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

export const DISASTER_CATEGORIES = [
  'fire',
  'flood',
  'accident',
  'infrastructure',
  'medical',
  'other',
] as const;

export type DisasterCategory = (typeof DISASTER_CATEGORIES)[number];

export const VERIFICATION_STATUSES = [
  'pending',
  'verified',
  'rejected',
  'flagged',
] as const;

export type VerificationStatus = (typeof VERIFICATION_STATUSES)[number];

export const AUTHORITY_STATUSES = [
  'pending',
  'dispatched',
  'resolved',
  'ignored',
] as const;

export type AuthorityStatus = (typeof AUTHORITY_STATUSES)[number];

// ---------------------------------------------------------------------------
// Zod schemas
// ---------------------------------------------------------------------------

/**
 * Validates the non-file fields of POST /disaster/report.
 * The image is handled separately via multer.
 */
export const disasterReportBodySchema = z.object({
  lat: z.coerce.number().min(-90).max(90),
  lon: z.coerce.number().min(-180).max(180),
  description: z.string().max(500).optional(),
});

/**
 * Validates query params for GET /disaster/feed.
 */
export const disasterFeedQuerySchema = z.object({
  category: z.enum(DISASTER_CATEGORIES).optional(),
  severity_min: z.coerce.number().int().min(1).max(5).optional(),
  since: z.string().datetime({ offset: true }).or(z.string().datetime()).optional(),
  ne_lat: z.coerce.number().min(-90).max(90).optional(),
  ne_lon: z.coerce.number().min(-180).max(180).optional(),
  sw_lat: z.coerce.number().min(-90).max(90).optional(),
  sw_lon: z.coerce.number().min(-180).max(180).optional(),
  page: z.coerce.number().int().min(1).default(1),
  limit: z.coerce.number().int().min(1).max(100).default(20),
});

/**
 * Validates PATCH /disaster/:id/status body.
 */
export const authorityStatusUpdateSchema = z.object({
  authority_status: z.enum(AUTHORITY_STATUSES),
});

/**
 * Validates GET /disaster/heatmap query.
 */
export const heatmapQuerySchema = z.object({
  precision: z.coerce.number().int().min(0).max(4).default(2),
  since: z.string().datetime({ offset: true }).or(z.string().datetime()).optional(),
  category: z.enum(DISASTER_CATEGORIES).optional(),
});

/**
 * Validates GET /disaster/stats query.
 */
export const statsQuerySchema = z.object({
  range: z.enum(['24h', '7d', '30d']).default('24h'),
});

// ---------------------------------------------------------------------------
// TypeScript interfaces
// ---------------------------------------------------------------------------

export interface DisasterReport {
  id: string;
  userId: string;
  lat: number;
  lon: number;
  imageUrl: string;
  imageHash: string;
  category: DisasterCategory;
  severityScore: number;
  llmConfidence: number;
  llmRawResponse: Record<string, unknown> | null;
  verificationStatus: VerificationStatus;
  rejectionReason: string | null;
  authorityStatus: AuthorityStatus;
  description: string | null;
  linkedSosId: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface GeminiVerificationResult {
  isReal: boolean;
  category: DisasterCategory;
  severity: number;
  confidence: number;
  reasoning: string;
  flags: string[];
}

export interface DisasterStats {
  totalReports: number;
  verified: number;
  rejected: number;
  flagged: number;
  pending: number;
  byCategory: Record<string, number>;
  bySeverity: Record<number, number>;
  sosCount: number;
}

export interface HeatmapPoint {
  lat: number;
  lon: number;
  count: number;
  avgSeverity: number;
}
